text/0000-incremental-compilation.md

```diff
@@ -0,0 +1,648 @@
+- MIR
+  - The MIR represents the type-checked statements in the body, in
+    simplified forms. It is described by [RFC #1211][1211]. As the MIR
+    is not fully implemented, this is a non-trivial dependency. We
+    could instead use the existing annotated HIR, however that would
+    require a larger effort in terms of porting and adapting data
+    structures to an incremental setting.  Using the MIR simplifies
+    things in this respect.
+- Object files
+  - This represents the final result of running LLVM. It may be that
+    the best strategy is to "cache" compiled code in the form of an
+    rlib that is progressively patched, or it may be easier to store
+    individual `.o` files that must be relinked (anyone who has worked
+    in a substantial C++ project can attest, however, that linking can
+    take a non-trivial amount of time).
+
+Of course, the key to any incremental design is to determine what must
+be changed. This can be encoded in a *dependency graph*. This graph
+connects the various bits of the HIR to the external products
+(signatures, MIR, and object files). It is of the utmost importance
+that this dependency graph is complete: if edges are missing, the
+result will be obscure errors where changes are not fully propagated,
+yielding inexplicable behavior at runtime. This RFC proposes an
+automatic scheme based on encapsulation.
+
+### Interaction with lints and compiler plugins
+
+Although rustc does not yet support compiler plugins through a stable
+interface, we have long planned to allow for custom lints, syntax
+extensions, and other sorts of plugins. It would be nice therefore to
+be able to accommodate such plugins in the design, so that their
+inputs can be tracked and accounted for as well.
+
+## Interaction with optimization
+
+It is important to clarify, though, that this design does not attempt
+to enable full optimizing for incremental compilation; indeed the two
+are somewhat at odds with one another, as full optimization may
+perform inlining and inter-function analysis, which can cause small
+edits in one function to affect the generated code of another. This
+situation is further exacerbated by the fact that LLVM does not
+provide any way to track these sorts of dependencies (e.g., one cannot
+even determine what inlining took place, though @dotdash suggested a
+clever trick of using llvm lifetime hints). Strategies for handling
+this are discussed in the [Optimization section](#optimization) below.
+
+# Detailed design
+
+We begin with a high-level execution plan, followed by sections that
+explore aspects of the plan in more detail. The high-level summary
+includes links to each of the other sections.
+
+## High-level execution plan
+
+Regardless of whether it is invoked in incremental compilation mode or
+not, the compiler will always parse and macro expand the entire crate,
+resulting in a HIR tree. Once we have a complete HIR tree, and if we
```

Comment:
IME most of the time spent in those cases is on later stages, processing the large amounts of code generated by the syntax extension.